File: app/src/main/java/com/dpis/module/PackageConfigValue.java
package com.dpis.module;

import com.dpis.module.viewport.DpiConfig;

import com.dpis.module.viewport.ViewportApplyMode;
import com.dpis.module.viewport.ViewportTargetSpec;
import com.dpis.module.viewport.ViewportTargetType;

import com.dpis.module.appconfig.WechatDpiConfig;

import java.util.Objects;

public final class PackageConfigValue {
    public static final PackageConfigValue EMPTY = new PackageConfigValue(
            ViewportTargetSpec.off(),
            ViewportTargetType.OFF,
            ViewportApplyMode.OFF,
            null,
            FontApplyMode.OFF,
            null,
            null,
            null,
            null);

    final ViewportTargetSpec viewportTargetSpec;
    final String viewportTargetType;
    final String viewportApplyMode;
    final Integer fontScalePercent;
    final String fontApplyMode;
    final String typefaceId;
    final String fontHookDomainsRaw;
    final Boolean dpisEnabled;
    final Integer wechatDpi;

    public PackageConfigValue(
            ViewportTargetSpec viewportTargetSpec,
            String viewportTargetType,
            String viewportApplyMode,
            Integer fontScalePercent,
            String fontApplyMode,
            String typefaceId,
            String fontHookDomainsRaw,
            Boolean dpisEnabled,
            Integer wechatDpi) {
        this.viewportTargetSpec = viewportTargetSpec != null
                ? viewportTargetSpec
                : ViewportTargetSpec.off();
        this.viewportTargetType = this.viewportTargetSpec.isEnabled()
                ? this.viewportTargetSpec.type()
                : ViewportTargetType.normalize(viewportTargetType);
        this.viewportApplyMode = ViewportApplyMode.normalize(viewportApplyMode);
        this.fontScalePercent = normalizeFontScalePercent(fontScalePercent);
        this.fontApplyMode = FontApplyMode.normalize(fontApplyMode);
        this.typefaceId = normalizeNullableString(typefaceId);
        this.fontHookDomainsRaw = normalizeNullableString(fontHookDomainsRaw);
        this.dpisEnabled = normalizeDpisEnabled(dpisEnabled);
        this.wechatDpi = WechatDpiConfig.normalize(wechatDpi);
    }

    public ViewportTargetSpec viewportTargetSpec() {
        return viewportTargetSpec;
    }

    public String viewportTargetType() {
        return viewportTargetType;
    }

    public String viewportApplyMode() {
        return viewportApplyMode;
    }

    public Integer fontScalePercent() {
        return fontScalePercent;
    }

    public String fontApplyMode() {
        return fontApplyMode;
    }

    public String typefaceId() {
        return typefaceId;
    }

    public String fontHookDomainsRaw() {
        return fontHookDomainsRaw;
    }

    public Boolean dpisEnabled() {
        return dpisEnabled;
    }

    public Integer wechatDpi() {
        return wechatDpi;
    }

    public boolean hasAnyValue() {
        return viewportTargetSpec.isEnabled()
                || !ViewportTargetType.OFF.equals(viewportTargetType)
                || isStoredViewportApplyMode(viewportApplyMode)
                || fontScalePercent != null
                || isStoredFontApplyMode(fontApplyMode)
                || typefaceId != null
                || fontHookDomainsRaw != null
                || dpisEnabled != null
                || wechatDpi != null;
    }

    private static String normalizeNullableString(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private static Integer normalizeFontScalePercent(Integer percent) {
        if (percent == null || percent < 50 || percent > 300) {
            return null;
        }
        return percent;
    }

    private static Boolean normalizeDpisEnabled(Boolean enabled) {
        if (enabled == null || enabled) {
            return null;
        }
        return Boolean.FALSE;
    }

    private static boolean isStoredViewportApplyMode(String mode) {
        return ViewportApplyMode.SYSTEM.equals(mode)
                || ViewportApplyMode.COMPAT.equals(mode);
    }

    private static boolean isStoredFontApplyMode(String mode) {
        return FontApplyMode.FIELD_REWRITE.equals(mode);
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) {
            return true;
        }
        if (!(object instanceof PackageConfigValue other)) {
            return false;
        }
        return viewportTargetSpec.equals(other.viewportTargetSpec)
                && viewportTargetType.equals(other.viewportTargetType)
                && viewportApplyMode.equals(other.viewportApplyMode)
                && Objects.equals(fontScalePercent, other.fontScalePercent)
                && fontApplyMode.equals(other.fontApplyMode)
                && Objects.equals(typefaceId, other.typefaceId)
                && Objects.equals(fontHookDomainsRaw, other.fontHookDomainsRaw)
                && Objects.equals(dpisEnabled, other.dpisEnabled)
                && Objects.equals(wechatDpi, other.wechatDpi);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                viewportTargetSpec,
                viewportTargetType,
                viewportApplyMode,
                fontScalePercent,
                fontApplyMode,
                typefaceId,
                fontHookDomainsRaw,
                dpisEnabled,
                wechatDpi);
    }
}

File: app/src/main/java/com/dpis/module/config/PackageConfigValue.kt
package com.dpis.module.config

import com.dpis.module.appconfig.WechatDpiConfig
import com.dpis.module.fonts.FontApplyMode
import com.dpis.module.viewport.ViewportApplyMode
import com.dpis.module.viewport.ViewportTargetSpec
import com.dpis.module.viewport.ViewportTargetType

class PackageConfigValue(
    viewportTargetSpec: ViewportTargetSpec?,
    viewportTargetType: String?,
    viewportApplyMode: String?,
    fontScalePercent: Int?,
    fontApplyMode: String?,
    typefaceId: String?,
    fontHookDomainsRaw: String?,
    dpisEnabled: Boolean?,
    wechatDpi: Int?,
) {
    @JvmField val viewportTargetSpec: ViewportTargetSpec =
        viewportTargetSpec ?: ViewportTargetSpec.off()
    @JvmField val fontScalePercent: Int? = fontScalePercent?.takeIf { it in 50..300 }
    @JvmField val typefaceId: String? = typefaceId?.trim()?.takeIf(String::isNotEmpty)
    @JvmField val fontHookDomainsRaw: String? = fontHookDomainsRaw?.trim()
    @JvmField val dpisEnabled: Boolean? = dpisEnabled?.takeIf { !it }
    @JvmField val wechatDpi: Int? = WechatDpiConfig.normalize(wechatDpi)
    @JvmField val viewportTargetType: String =
        if (this.viewportTargetSpec.isEnabled()) this.viewportTargetSpec.type()
        else ViewportTargetType.normalize(viewportTargetType)
    @JvmField val viewportApplyMode: String = ViewportApplyMode.normalize(viewportApplyMode)
    @JvmField val fontApplyMode: String = FontApplyMode.normalize(fontApplyMode)

    fun viewportTargetSpec(): ViewportTargetSpec = viewportTargetSpec
    fun viewportTargetType(): String = viewportTargetType
    fun viewportApplyMode(): String = viewportApplyMode
    fun fontScalePercent(): Int? = fontScalePercent
    fun fontApplyMode(): String = fontApplyMode
    fun typefaceId(): String? = typefaceId
    fun fontHookDomainsRaw(): String? = fontHookDomainsRaw
    fun dpisEnabled(): Boolean? = dpisEnabled
    fun wechatDpi(): Int? = wechatDpi

    fun hasAnyValue(): Boolean =
        viewportTargetSpec().isEnabled() ||
            viewportTargetType != ViewportTargetType.OFF ||
            viewportApplyMode == ViewportApplyMode.SYSTEM ||
            viewportApplyMode == ViewportApplyMode.COMPAT ||
            fontScalePercent != null ||
            fontApplyMode == FontApplyMode.FIELD_REWRITE ||
            typefaceId() != null ||
            fontHookDomainsRaw() != null ||
            dpisEnabled() != null ||
            wechatDpi() != null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PackageConfigValue) return false
        return viewportTargetSpec() == other.viewportTargetSpec() &&
            viewportTargetType == other.viewportTargetType &&
            viewportApplyMode == other.viewportApplyMode &&
            fontScalePercent == other.fontScalePercent &&
            fontApplyMode == other.fontApplyMode &&
            typefaceId() == other.typefaceId() &&
            fontHookDomainsRaw() == other.fontHookDomainsRaw() &&
            dpisEnabled() == other.dpisEnabled() &&
            wechatDpi() == other.wechatDpi()
    }

    override fun hashCode(): Int = arrayOf(
        viewportTargetSpec(), viewportTargetType, viewportApplyMode, fontScalePercent,
        fontApplyMode, typefaceId(), fontHookDomainsRaw(), dpisEnabled(), wechatDpi(),
    ).contentHashCode()

    companion object {
        @JvmField val EMPTY = PackageConfigValue(
            ViewportTargetSpec.off(),
            ViewportTargetType.OFF,
            ViewportApplyMode.OFF,
            null,
            FontApplyMode.OFF,
            null,
            null,
            null,
            null,
        )
    }
}
